ESPS/general/src/window/window.1
.\" Copyright (c) 1987,1988,1990 Entropic Speech, Inc.; All rights reserved
.\" @(#)window.1	1.8 1.8 ESI
.TH WINDOW 1\-ESPS 3/16/90
.ds ]W "\fI\s+4\ze\h'0.05'e\s-4\v'-0.4m'\fP\(*p\v'0.4m'\ Entropic Speech, Inc.
.SH "NAME"
window \- windows sampled data in FEA records 
.SH "SYNOPSIS"
.B window
[
.BI \-P " params"
] [
.BI \-f " sd_field"
[
.BI \-f " power_field"
] [
.BI \-r " range"
] [
.BI \-l " window_len"
] [
.BI \-w " window_type"
] [
.BI \-x " debug_level"
]
.I "input output"
.SH "DESCRIPTION"
.PP
.I Window
accepts a FEA file \fIinput\fP containing a vector sampled-data field
in each record.  It produces a FEA file \fIoutput\fP with records
containing the results of applying a fixed window to the input
sampled-data field.  Each field of sampled data may be thought of as a
separate frame of data, such as might be produced by \fIframe\fP
(1\-\s-1ESPS\s+1).  The default name for the sampled data field in
\fIinput\fP is \fIsd\fP, and the default name for the window field in
\fIoutput\fP is \fIwind_sd\fP.  Both defaults can be changed by means
of the \fB\-f\fP option.  If the input is tagged or segment labelled,
the relevant information is copied over to the output.  If \fIinput\fP
is "\-", standard input is used for the input file.  If \fIoutput\fP
is "\-", standard input is used for the output file.
.PP
The window is computed using \fIwindow\fP (3\-\s-1ESPS\s+1).  
The fixed window length \fIwindow_len\fP must not be larger than the size
of the sampled data field.  If it is smaller, zeros are filled to the 
end of the field.  All of the first \fIwindow_len\fP elements in 
each input field are processed.  If \fIinput\fP is segment-labelled, 
\fIwindow\fP warns if any of the \fIsegment_length\fP values are 
less than \fIwindow_len\fP.  (This may be intentional, so it 
would be too verbose to identify each such frame.  They may, however,
be identified using \fIselect\fP (1\-\s-1ESPS\s+1).)  
.PP
Note that windowed FEA sampled data records can also be produced 
by \fIframe\fP (1\-\s-1ESPS\s+1).  \fIWindow\fP is intended for 
occasions when it is appropriate to defer windowing because of 
intermediate processing.  
.SH OPTIONS
.PP
The following options are supported:
.TP
.BI \-P " param" " \fR[params]\fP"
Specifies the name of the parameter file.
.TP
.BI \-r " start" : "last" "\fR [1:(last in file)]"
.TP
.BI \-r " start" :+ "nan"
In the first form, a pair of unsigned integers specifies the range
of records to be processed.
Either
.I start
or
.I last
may be omitted; then the default value is used.
If
.IR last " = " start " + " nan,
the second form (with the plus sign) specifies the same range as the first.
The \fB\-r\fP overrides the values of \fIstart\fP and \fInan\fP from 
the parameter file.  
.TP
.BI \-f " field_name" 
If this option is used once, it specifies the name of the sampled 
data field in \fIinput\fP.  If it is used twice, the second time
it specifies the name of the window field in \fIoutput\fP.  
The default names for these fields are "sd" and "wind_sd", respectively. 
.TP
.BI \-l " window_len" "\fR [0]"
Specifies the window length.  If the option is omitted, the
parameter file is consulted.  A value of 0 (from either the option or the
parameter file) indicates that the window length is the size of the sampled
data field in \fIinput\fP.  This is also the default value in case
.I window_len
is not specified either with the
.B \-l
option or in the parameter file.
.TP
.BI \-w " window_type" "\fR[RECT]"
The name of the data window to apply to the data in each frame before
computing reflection coefficients.  If the option is omitted, the parameter
file is consulted, and if no value is found there, the default used is a
rectangular window with amplitude one.  Possible window types include
rectangular ("RECT"), Hamming ("HAMMING"), and triangular ("TRIANG"); see
the window(3-ESPSsp) manual page for the complete list of supported
window types.
.IP
If the \fB\-w\fP specified
value is not a supported window type,
it is assumed to be the name of an ascii file containing
weighting function values. If no file by this name exists,
\fIwindow\fP (1\-\s-1ESPS\s+1) warns and exits.
Otherwise the file is read, the weighting values are saves, and the number
of weighting values is counted (\fIwindow_len\fP). 
In this case, the \fIwindow_len\fP value from either the \fB\-l\fP option
or the specified \fIparams\fP file is ignored.
The same rules 
about \fIwindow_len\fP as described above in \fBDESCRIPTION\fP
are in effect.
.SH "ESPS PARAMETERS"
.PP
The parameter file does not have to be present,
since all the parameters have default values.
The following parameters are read, if present, from the parameter file:
.TP
.I "sd_field_name \- string"
.IP
This is the name of the sampled data field in \fIinput\fP.
The default is "sd".  A paramter file value (if present) is overidden
by the first use of the \fB\-f\fP option.  
.TP
.I "wind_field_name \- string"
.IP
This is the name of the windowed sampled data field in \fIoutput\fP.  The
default is "wind_sd".  A paramter file value (if present) is overidden
by the second use of the \fB\-f\fP option.
.TP
.I "start \- integer"
.IP
This is the first record of \fIinput\fP to process.
The default is 1.
It is not read if the \fB\-r\fP option is used.  
.TP
.I "nan \- integer"
.IP
This is the number of records to process.  It is not read if the \fB\-r\fP
option is used.
A value of zero means all subsequent records in the file;
this is the default.
.TP
.I "window_len - integer"
.IP
The window size. This parameter is not read if the
.B \-l
option is specified.  A value of 0 indicates that the window length is the
size of the sampled data field in \fIinput\fP; this is also the default
value in case
.I window_len
is specified neither with the
.B \-l
option nor in the parameter file.
.TP
.I "window_type \- string"
The data window to apply to the data.
This parameter is not read if the command-line option
.B \-w
is specified.  If the option is omitted and if no value is found in
the parameter file, the default used is "RECT", for a rectangular
window with amplitude one.  
If the specified value is not a supported window type,
it is assumed to be a user-defined window function.
See the discussion of the \fB-w\fP option.
.PP
Remember that command line option values override parameter file
values.
.SH "ESPS COMMON"
.PP
ESPS Common processing may be disabled by setting the environment variable
USE_ESPS_COMMON to "off".  The default ESPS Common file is .espscom 
in the user's home directory.  This may be overidden by setting
the environment variable ESPSCOM to the desired path.  User feedback of
Common processing is determined by the environment variable ESPS_VERBOSE,
with 0 causing no feedback and increasing levels causing increasingly
detailed feedback.  If ESPS_VERBOSE is not defined, a default value of 3 is
assumed.
.PP
ESPS Common is not processed by 
.I window
if \fIinput\fP is 
standard input.  Otherwise, provided that the Common file is newer 
than the parameter file, and provided that the \fIfilename\fP entry in 
Common matches \fIinput\fP, the Common values for \fIstart\fP and 
\fInan\fP override those that may be present in the parameter file. 
.PP
The following items are written into the ESPS Common file 
provided that \fIoutput\fP is not <stdout>.  
.IP
.I "start \- integer"
.IP
The starting point from the input file.
.sp
.I "nan \- integer"
.IP
The number of points in the selected range.
.sp
.I "prog \- string"
.IP
This is the name of the program 
.RI ( window
in this case).
.sp
.I "filename \- string"
.IP
The name of the input file \fIinput\fP.
.SH ESPS HEADERS
.PP
The \fIoutput\fP header is a new FEA file header.  If it exists in the
input header, the generic header item \fIsrc_sf\fP is copied from
input to output.  Following the convention from \fIframe\fP
(1\-\s-1ESPS\s+1), this generic is intended to contain the sampling
rate of the original sampled data.  The generic header item
\fIstart_time\fP is written with a value computed by taking the
\fIstart_time\fP value from the header of the input file (or zero, if
such a header item doesn't exist) and adding to it the relative time
from the first record in the file to the first record processed.  The
computation of \fIstart_time\fP depends on the value of the generic
header item \fIrecord_freq\fP in the input file.  If this item is not
present, \fIstart_time\fP is just copied from the input file to the
output.
.PP
The items \fIstart\fP and \fInan\fP are rewritten (if they already
exist) to contain the starting record number and number of records
processed.  Generic header items are added for the \fIwindow_len\fP
and \fIwindow_type\fP.  As usual, the command line is added as a
comment and the header of \fIinput\fP is added as a source file to
\fIoutput\fP.  
.SH "FUTURE CHANGES"
.PP
Control over the type of the output windowed sampled data field.  
.SH "SEE ALSO"
.PP
\fIframe\fP (1\-\s-1ESPS\s+1), window (3\-\s-1ESPS\s+1), \fIpwr\fP
(1\-\s-1ESPS\s+1), FEA (5\-\s-1ESPS\s+1), \fImake_sd\fP
(1\-\s-1ESPS\s+1), \fIplotfield\fP (1\-\s-1ESPS\s+1)
.SH "WARNINGS AND DIAGNOSTICS"
.PP
.I window
will exit with an error message if any of the 
following are true: 
.I input
does not exist or is not an ESPS FEA file; 
the sampled-data field does not exist in
.I input;
the window field already exists in
.I input;
the \fIwindow_len\fP is larger than the size of the \fIinput\fP sampled
data field.
.SH "BUGS"
.PP
None known.  
.SH "AUTHOR"
.PP
Manual page and program by John Shore. 



ESPS/general/src/window/window.c
/*----------------------------------------------------------------------+
|									|
|   This material contains proprietary software of Entropic Speech,	|
|   Inc.  Any reproduction, distribution, or publication without the	|
|   prior written permission of Entropic Speech, Inc. is strictly	|
|   prohibited.  Any public distribution of copies of this work		|
|   authorized in writing by Entropic Speech, Inc. must bear the	|
|   notice								|
|									|
|   "Copyright (c) 1988, 1990 Entropic Speech, Inc.  All rights reserved."
|									|
+-----------------------------------------------------------------------+
|									|
|  Program: window.c							|
|									|
|  This program windows the data in sampled-data records in a FEA	|
|  file.  The output is contains copies of the input records with a	|
|  field containing the windowed data.					|
|									|
|  John Shore, Entropic Speech, Inc.			         	|
|  based on power.c by Rodney Johnson					|
|  User-defined window added by David Burton				|
+----------------------------------------------------------------------*/
#ifndef lint
    static char *sccs_id = "@(#)window.c	1.10	8/31/95	ESI";
#endif
#define VERSION "1.10"
#define DATE "8/31/95"

#include <stdio.h>
#include <esps/esps.h>
#include <esps/unix.h>
#include <esps/fea.h>
#include <esps/window.h>

#define REQUIRE(test,text) {if (!(test)) {(void) fprintf(stderr, \
"%s: %s - exiting\n", ProgName, text); exit(1);}}

#define SYNTAX USAGE("window [-f sd_field [-f winsd_field]] [-l window_length]\n [-r range] [-x debug_level] [-w window_type] [-P params]\n input.fea output.fea") ;

#define WT_PREFIX "WT_"


void		pr_farray();
char		*get_cmd_line();
void		lrange_switch();
char            *type_convert();
char            *atoarray();
char	*ProgName = "window";
char	*Version = VERSION;
char	*Date = DATE;
int	debug_level = 0;	/* debug level, global for library*/

/*array of field names to copy over from input to output (these are for 
 *copying over the segment labelling*/

char   *seg_fields[] = {"source_file", "segment_start", "segment_length", NULL};


/*
 * MAIN PROGRAM
 */

main(argc, argv)
    int  argc;
    char **argv;
{
    extern int	    optind;	/* for use of getopt() */
    extern char	    *optarg;	/* for use of getopt() */
    int		    ch;		/* command-line option letter */

    int		    fflag = 0;	/* -f option specified 0, 1, or 2 times? */
    char	    *sd_name;	/* field name for input sampled data */
    int		    sd_type;	/* data type of sampled-data field */
    char	    *wsd_name;	/* field name for windowed sampled-data */
    char	    *window_type = NULL;	
				/* name of type of window to apply to data */
    char	    *pref_w_type;	
				/* window type name with added prefix */
    int		    win;    	/* window type code */
    int		    wflag = 0;	/* flag for window option*/
    int		    rflag = NO;	/* -r option specified? */
    int		    lflag = NO; /* -l option specified? */
    char	    *rrange;	/* arguments of -r option */
    long	    start_rec;	/* starting record number */
    long	    end_rec;	/* ending record number */
    long	    num_recs;	/* number of records to read
				    (0 means all up to end of file) */
    long	    num_read;	/* number of records actually read */

    char	    *param_name = NULL;
				/* parameter file name */

    char	    *iname;	/* input file name */
    FILE	    *ifile;	/* input stream */
    FILE   *instrm = stdin;     /* for ASCII data*/
    struct header   *ihd;	/* input file header */
    struct fea_data *irec;	/* input record */
    char	    *sd_ptr;	/* pointer (untyped) to sampled-data
				    field in input record */
    long	    *frame_len;	/* pointer to frame size
				    in input header or record */
    long	    size;	/* size of input sampled-data field */
    long	    win_len;   	/* size of window to apply */
    float	    *data;	/* input data as float */
    char	    *oname;	/* output file name */
    FILE	    *ofile;	/* output stream */
    struct header   *ohd;	/* output file header */
    struct fea_data *orec;	/* output record */
    float	    *wsd_ptr;	/* pointer to windowed 
				    sampled data field in output record */

    int		    i;		/* loop index */
    int		    seglenwarn = 0;	
				/* warning flag for segment lengths less
				    than window length*/
    short           tagged=0;	/* flag to indicate input is tagged */
    short           seglab=0;	/* to indicate segmenta labelled input */
    short           file_ind;	/* index for location of source_file field */
    extern char
	*window_types[];	/* standard window type names */

    double          *udefined;  /*pointer for user defined window coeffs*/

/*
 * Parse command-line options.
 */

    while ((ch = getopt(argc, argv, "f:r:x:P:l:w:")) != EOF)
        switch (ch)
	{
	case 'f':
	    switch (fflag)
    	    {
	    case 0:
		sd_name	= optarg;
		fflag++;
		break;
	    case 1:
		wsd_name = optarg;
		fflag++;
		break;
	    default:		
		Fprintf(stderr, 
		    "%s: -f option may be specified at most twice.\n",
		    ProgName);
		exit(1);
	    }
	    break;
	case 'r':
	    rflag = YES;
	    rrange = optarg;
	    break;
	case 'x':
	    debug_level = atoi(optarg);
	    break;
	case 'P':
	    param_name = optarg;
	    break;
	case 'l':
	    win_len = atoi(optarg);
	    lflag = YES;
	    break;
	case 'w':
	    window_type = optarg;
	    wflag++;
	    break;
	default:
	    SYNTAX
	    break;
	}

/*
 * Process file names and open files.
 */

    if (argc - optind > 2)
    {
	Fprintf(stderr,
	    "%s: too many file names specified.\n", ProgName);
	SYNTAX
    }
    if (argc - optind < 2)
    {
	Fprintf(stderr,
	    "%s: too few file names specified.\n", ProgName);
	SYNTAX
    }

    iname = eopen(ProgName,
	    argv[optind++], "r", FT_FEA, NONE, &ihd, &ifile);

    oname = eopen(ProgName,
	    argv[optind++], "w", NONE, NONE, &ohd, &ofile);

    if (debug_level)
	Fprintf(stderr, "Input file: %s\nOutput file: %s\n",
	    iname, oname);

/*
 * Get parameter values.
 */

    if (ifile != stdin)
	(void) read_params(param_name, SC_CHECK_FILE, iname);
    else
        (void) read_params(param_name, SC_NOCOMMON, iname);

    if (fflag < 1)
	sd_name =
	    (symtype("sd_field_name") != ST_UNDEF)
	    ? getsym_s("sd_field_name")
	    : "sd";

    if (fflag < 2)
	wsd_name =
	    (symtype("wind_field_name") != ST_UNDEF)
	    ? getsym_s("wind_field_name")
	    : "wind_sd";

    if (debug_level)
	Fprintf(stderr,
	    "sd_field_name: \"%s\".  wind_field_name: \"%s\".\n",
	    sd_name, wsd_name);

    sd_type = get_fea_type(sd_name, ihd);
    REQUIRE(sd_type != UNDEF, "sampled-data field not in input file");

    REQUIRE(!is_field_complex(ihd, sd_name), 
	    "sorry, can't deal with complex data"); 

    if (rflag)
    {
	start_rec = 1;
	end_rec = LONG_MAX;
	lrange_switch(rrange, &start_rec, &end_rec, 0);
	num_recs =
	    (end_rec != LONG_MAX)
	    ? end_rec - start_rec + 1
	    : 0;
    }
    else
    {
	start_rec =
	    (symtype("start") != ST_UNDEF)
	    ? getsym_i("start")
	    : 1;
	num_recs =
	    (symtype("nan") != ST_UNDEF)
	    ? getsym_i("nan")
	    : 0;
	end_rec = (num_recs != 0)
	    ? start_rec - 1 + num_recs
	    : LONG_MAX;
    }

    if (!lflag) 
	win_len =
	    (symtype("window_len") != ST_UNDEF)
	    ? getsym_i("window_len")
	    : 0;
	
    if (debug_level)
	Fprintf(stderr, "start_rec: %ld.  end_rec: %ld.  num_recs: %ld.\n",
	    start_rec, end_rec, num_recs);

    if (debug_level)
	Fprintf(stderr, "window_len: %ld.\n", win_len);

    REQUIRE(start_rec >= 1, "can't start before beginning of file");
    REQUIRE(end_rec >= start_rec, "empty range of records specified");

    if (!wflag)
	window_type =
	    (symtype("window_type") != ST_UNDEF)
	    ? getsym_s("window_type")
	    : "HAMMING";

    pref_w_type = 
	malloc((unsigned)(strlen(WT_PREFIX) + strlen(window_type) + 1));
    REQUIRE(pref_w_type, "can't allocate space for window type name");
    (void) strcpy(pref_w_type, WT_PREFIX);
    (void) strcat(pref_w_type, window_type);

    win = lin_search(window_types, pref_w_type);

    if(win == -1){
      Fprintf(stderr, "window: Window type is not recognized.\n\t Assume type is name of ASCII file (or stdin) containing weight coefficients\n");
      /*
       * open file with coefficients
       */
      if (strcmp (window_type, "-") == 0)
        window_type = "<stdin>";
      else 
        TRYOPEN(argv[0], window_type, "r", instrm);

      if(strcmp(window_type, iname) == 0){
	Fprintf(stderr, "window: input window coefficients and signal data cannot both come from standard input - exiting.\n");
	exit(1);
      }

      udefined = (double *) atoarray(instrm, DOUBLE, &win_len, (double *)NULL);
      win = WT_ARB;
     }

    if (debug_level)
	Fprintf(stderr, "%s: window_type = %s, win = %d\n",
	    ProgName, window_type, win);

    irec = allo_fea_rec(ihd);
    REQUIRE(irec != NULL, "can't allocate memory for input record");

    sd_ptr = get_fea_ptr(irec, sd_name, ihd);

    REQUIRE(sd_ptr, "can't locate sampled-data field in input record");

    size = get_fea_siz(sd_name, ihd, (short *) NULL, (long **) NULL);
    REQUIRE(size != -1,
	"bad sampled-data field definition in input file");
    REQUIRE(win_len <= size, "window_len larger than sd field size");

    if (ihd->hd.fea->segment_labeled)
    {
	frame_len = (long *) get_fea_ptr(irec, "segment_length", ihd);
	REQUIRE(frame_len, "no segment_length field in input file header");

	if (debug_level)
	    Fprintf(stderr, "Input file segment_labeled.\n");

	file_ind = lin_search2(ihd->hd.fea->names,"source_file");

	set_seg_lab(ohd, ihd->hd.fea->enums[file_ind]);

	seglab = 1;

    }
    else
    {
	frame_len = &size;

	if (debug_level)
	    Fprintf(stderr,
		"Input file not segment_labeled.  Frame length: %ld.\n",
		*frame_len);
    }

    /* 0 input for window_len (default) means size of sd field*/

    if (win_len == 0) 
	win_len = size;
/*
 * Create and write output-file header
 */

    ohd = new_header(FT_FEA);

    if (ihd->common.tag == YES) {
	ohd->common.tag = YES;
	ohd->variable.refer = ihd->variable.refer;
	tagged = 1; 
	ohd->variable.refhd = ihd->variable.refhd;
    }

    (void) strcpy(ohd->common.prog, ProgName);
    (void) strcpy(ohd->common.vers, Version);
    (void) strcpy(ohd->common.progdate, Date);

    add_source_file(ohd, iname, ihd);
    add_comment(ohd, get_cmd_line(argc, argv));

    *add_genhd_e( "window_type", (short *) NULL, 1, window_types, ohd) = win;

    REQUIRE( add_fea_fld(wsd_name, size, (short) 1, (long *) NULL,
			    FLOAT, (char **) NULL, ohd) != -1,
		"can't create wind_field_name field in output file header" );

    *add_genhd_l( "window_len", (long *) NULL, 1, ohd) = win_len;

    *add_genhd_l("nan", (long *) NULL, 1, ohd) = num_recs;
    *add_genhd_l("start", (long *) NULL, 1, ohd) = start_rec;

    (void) copy_genhd(ohd, ihd, "src_sf");

    update_waves_gen(ihd, ohd, (float) start_rec, 1.0);

    if (debug_level)
	Fprintf(stderr, "Writing output header to file.\n");

    write_header(ohd, ofile);

/*
 * Allocate records and set up pointers
 */

    orec = allo_fea_rec(ohd);
    REQUIRE(orec != NULL, "can't allocate memory for output record");

    wsd_ptr = (float *) get_fea_ptr(orec, wsd_name, ohd);
    REQUIRE(wsd_ptr, "can't locate wind_samp_data field in output record");

    data = calloc_f((unsigned) size);
    REQUIRE(data,
	"can't allocate space for array of data as floats");

    for (i = 0; i < size; i++) 
	wsd_ptr[i] = 0.0;
/*
 * Main read-write loop
 */

    num_read = start_rec - 1,

    fea_skiprec(ifile, num_read, ihd);

    for (   ;
	    num_read < end_rec && get_fea_rec(irec, ihd, ifile) != EOF;
	    num_read++
	)
    {
	if (debug_level > 2)
	    Fprintf(stderr, "Record number %ld read.\n", num_read + 1);

	if (tagged)
	  orec->tag = irec->tag;
	else if (seglab)
	  copy_fea_rec(irec, ihd, orec, ohd, seg_fields, (short **) NULL);

	(void) type_convert(win_len, (char *) sd_ptr, sd_type, 
			    (char *) data, FLOAT, (void (*)())NULL);

    	if (debug_level > 3) pr_farray("sampled data input", 
	    win_len, data);

	if (*frame_len < win_len) 
	    seglenwarn++;

	/* window it, with results going directly to output record*/
	if(win != WT_ARB){
	   (void) window(win_len, data, wsd_ptr, win, (double *) NULL);
	   (void)add_genhd_c("window_data_in", window_type, 1, ohd);
	 }
	else
	   (void)window(win_len, data, wsd_ptr, win, udefined);

    	if (debug_level > 3) pr_farray("windowed sampled data", 
	    win_len, wsd_ptr);

	put_fea_rec(orec, ohd, ofile);

    }

    if (seglenwarn) Fprintf(stderr, 
	"%s: WARNING - some segment_length values less than window size\n",
	    ProgName);

    if (num_read < end_rec && num_recs != 0)
	Fprintf(stderr, "window: only %ld records read.\n",
		num_read - start_rec + 1);

/*
 * Write common
 */

    if (ofile != stdout)
    {
	REQUIRE(putsym_i("start", (int) start_rec) != -1,
		"can't write \"start\" to Common");
	REQUIRE(putsym_i("nan", (int) num_recs) != -1,
		"can't write \"nan\" to Common");
	REQUIRE(putsym_s("prog", ProgName) != -1,
		"can't write \"prog\" to Common");
	REQUIRE(putsym_s("filename", iname) != -1,
		"can't write \"filename\" to Common");
    }

    exit(0);
    /*NOTREACHED*/
}


/*
 * For debug printout of float arrays
 */

void pr_farray(text, n, arr)
    char    *text;
    long    n;
    float  *arr;
{
    int	    i;

    Fprintf(stderr, "%s: %s -- %d points:\n", ProgName, text, n);
    for (i = 0; i < n; i++)
    {
	Fprintf(stderr, "%f ", arr[i]);
	if (i%5 == 4 || i == n - 1) Fprintf(stderr,  "\n");
    }
}

